exchanges/kline/request.go

```diff
@@ -0,0 +1,201 @@
+	errNoTimeSeriesDataToConvert = errors.New("no time series data to convert")
+
+	// PartialCandle is string flag for when the most recent candle is partially
+	// formed.
+	PartialCandle = "Partial Candle"
+)
+
+// Request is a helper to request and convert time series to a required candle
+// interval.
+type Request struct {
+	// Exchange refers to the exchange name
+	Exchange string
+	// Pair refers to the currency pair
+	Pair currency.Pair
+	// RequestFormatted refers to the currency pair formatted by the exchange
+	// asset for outbound requests
+	RequestFormatted currency.Pair
+	// Asset refers to the asset type
+	Asset asset.Item
+	// ExchangeInterval refers to the interval that is used to construct the
+	// client required interval, this will be less than or equal to the client
+	// required interval.
+	ExchangeInterval Interval
+	// ClientRequired refers to the clients' actual required interval
+	// needed.
+	ClientRequired Interval
+	// Start is the start time aligned to UTC and to the Required interval candle
+	Start time.Time
+	// End is the end time aligned to UTC and to the Required interval candle
+	End time.Time
+	// PartialCandle defines when a request's end time interval goes beyond
+	// current time it potentially has a partially formed candle.
+	PartialCandle bool
+}
+
+// CreateKlineRequest generates a `Request` type for interval conversions
+// supported by an exchange.
+func CreateKlineRequest(name string, pair, formatted currency.Pair, a asset.Item, clientRequired, exchangeInterval Interval, start, end time.Time) (*Request, error) {
+	if name == "" {
+		return nil, ErrUnsetName
+	}
+	if pair.IsEmpty() {
+		return nil, currency.ErrCurrencyPairEmpty
+	}
+	if formatted.IsEmpty() {
+		return nil, currency.ErrCurrencyPairEmpty
+	}
+	if !a.IsValid() {
+		return nil, asset.ErrNotSupported
+	}
+	if clientRequired == 0 {
+		return nil, fmt.Errorf("client required %w", ErrInvalidInterval)
+	}
+	if exchangeInterval == 0 {
+		return nil, fmt.Errorf("exchange interval %w", ErrInvalidInterval)
+	}
+	err := common.StartEndTimeCheck(start, end)
+	if err != nil {
+		return nil, err
+	}
+
+	// Force UTC alignment
+	start = start.UTC()
+	end = end.UTC()
+
+	// Force alignment to required interval which is the higher time value e.g.
+	// 1hr required as opposed to 1min request/outbound interval used to
+	// construct the higher time value candle. This is to make sure there are
+	// minimal missing candles which is used to create the bigger candle.
+	start = start.Truncate(clientRequired.Duration())
+
+	// Strip monotonic clock reading for comparison
+	end = end.Round(0)
+
+	endTrunc := end.Truncate(clientRequired.Duration())
+	// Check to see if truncation moves end time and if so we want to make sure
+	// the candle period is included on the end.
+	if !endTrunc.Equal(end) {
+		end = endTrunc.Add(clientRequired.Duration())
+	}
+	return &Request{name, pair, formatted, a, exchangeInterval, clientRequired, start, end, end.After(time.Now())}, nil
```

Comment:
Think some funky stuff has been introduced. The time of writing this the time is 17/01/2023 14:03:00 AEST, so candles being returned don't exist yet
```
 go run . gethistoriccandlesextended --exchange="binance" --pair="btc-usdt" --asset="spot" -i="360" --start="2023-01-17 13:01:00" --end="2023-01-18 02:30:00"
{
 "exchange": "Binance",
 "pair": {
  "delimiter": "-",
  "base": "btc",
  "quote": "usdt"
 },
 "start": "2023-01-17 02:01:00 UTC",
 "end": "2023-01-17 15:30:00 UTC",
 "interval": "6m",
 "candle": [
  {
   "time": "2023-01-17 02:00:00 UTC",
   "low": 21030.5,
   "high": 21064.56,
   "open": 21039.96,
   "close": 21032.93,
   "volume": 770.74279
  },
  {
   "time": "2023-01-17 02:06:00 UTC",
   "low": 21017.44,
   "high": 21052.22,
   "open": 21031.61,
   "close": 21029.95,
   "volume": 639.48987
  },
  {
   "time": "2023-01-17 02:12:00 UTC",
   "low": 20986.34,
   "high": 21034.48,
   "open": 21029,
   "close": 21001.15,
   "volume": 734.79445
  },
  {
   "time": "2023-01-17 02:18:00 UTC",
   "low": 20998.81,
   "high": 21057,
   "open": 21000.14,
   "close": 21039.7,
   "volume": 843.99148
  },
  {
   "time": "2023-01-17 02:24:00 UTC",
   "low": 21028.44,
   "high": 21060.62,
   "open": 21039.7,
   "close": 21038.85,
   "volume": 547.47124
  },
  {
   "time": "2023-01-17 02:30:00 UTC",
   "low": 21033.56,
   "high": 21059.13,
   "open": 21038.85,
   "close": 21055.34,
   "volume": 571.4322
  },
  {
   "time": "2023-01-17 02:36:00 UTC",
   "low": 21044.7,
   "high": 21064.63,
   "open": 21054.47,
   "close": 21051.69,
   "volume": 462.54401
  },
  {
   "time": "2023-01-17 02:42:00 UTC",
   "low": 21041.35,
   "high": 21088.04,
   "open": 21050.68,
   "close": 21083,
   "volume": 752.00359
  },
  {
   "time": "2023-01-17 02:48:00 UTC",
   "low": 21071.09,
   "high": 21095.72,
   "open": 21083,
   "close": 21095.28,
   "volume": 687.41426
  },
  {
   "time": "2023-01-17 02:54:00 UTC",
   "low": 21070.93,
   "high": 21110.86,
   "open": 21094.16,
   "close": 21102.13,
   "volume": 771.7192600000001
  },
  {
   "time": "2023-01-17 03:00:00 UTC",
   "high": 21113.9,
   "open": 21102.13,
   "volume": 471.16115
  },
  {
   "time": "2023-01-17 03:06:00 UTC"
  },
  {
   "time": "2023-01-17 03:12:00 UTC"
  },
  {
   "time": "2023-01-17 03:18:00 UTC"
  },
  {
   "time": "2023-01-17 03:24:00 UTC"
  },
  {
   "time": "2023-01-17 03:30:00 UTC"
  },
  {
   "time": "2023-01-17 03:36:00 UTC"
  },
  {
   "time": "2023-01-17 03:42:00 UTC"
  },
  {
   "time": "2023-01-17 03:48:00 UTC"
  },
  {
   "time": "2023-01-17 03:54:00 UTC"
  },
  {
   "time": "2023-01-17 04:00:00 UTC"
  },
  {
   "time": "2023-01-17 04:06:00 UTC"
  },
  {
   "time": "2023-01-17 04:12:00 UTC"
  },
  {
   "time": "2023-01-17 04:18:00 UTC"
  },
  {
   "time": "2023-01-17 04:24:00 UTC"
  },
  {
   "time": "2023-01-17 04:30:00 UTC"
  },
  {
   "time": "2023-01-17 04:36:00 UTC"
  },
  {
   "time": "2023-01-17 04:42:00 UTC"
  },
  {
   "time": "2023-01-17 04:48:00 UTC"
  },
  {
   "time": "2023-01-17 04:54:00 UTC"
  },
  {
   "time": "2023-01-17 05:00:00 UTC"
  },
  {
   "time": "2023-01-17 05:06:00 UTC"
  },
  {
   "time": "2023-01-17 05:12:00 UTC"
  },
  {
   "time": "2023-01-17 05:18:00 UTC"
  },
  {
   "time": "2023-01-17 05:24:00 UTC"
  },
  {
   "time": "2023-01-17 05:30:00 UTC"
  },
  {
   "time": "2023-01-17 05:36:00 UTC"
  },
  {
   "time": "2023-01-17 05:42:00 UTC"
  },
  {
   "time": "2023-01-17 05:48:00 UTC"
  },
  {
   "time": "2023-01-17 05:54:00 UTC"
  },
  {
   "time": "2023-01-17 06:00:00 UTC"
  },
  {
   "time": "2023-01-17 06:06:00 UTC"
  },
  {
   "time": "2023-01-17 06:12:00 UTC"
  },
  {
   "time": "2023-01-17 06:18:00 UTC"
  },
  {
   "time": "2023-01-17 06:24:00 UTC"
  },
  {
   "time": "2023-01-17 06:30:00 UTC"
  },
  {
   "time": "2023-01-17 06:36:00 UTC"
  },
  {
   "time": "2023-01-17 06:42:00 UTC"
  },
  {
   "time": "2023-01-17 06:48:00 UTC"
  },
  {
   "time": "2023-01-17 06:54:00 UTC"
  },
  {
   "time": "2023-01-17 07:00:00 UTC"
  },
  {
   "time": "2023-01-17 07:06:00 UTC"
  },
  {
   "time": "2023-01-17 07:12:00 UTC"
  },
  {
   "time": "2023-01-17 07:18:00 UTC"
  },
  {
   "time": "2023-01-17 07:24:00 UTC"
  },
  {
   "time": "2023-01-17 07:30:00 UTC"
  },
  {
   "time": "2023-01-17 07:36:00 UTC"
  },
  {
   "time": "2023-01-17 07:42:00 UTC"
  },
  {
   "time": "2023-01-17 07:48:00 UTC"
  },
  {
   "time": "2023-01-17 07:54:00 UTC"
  },
  {
   "time": "2023-01-17 08:00:00 UTC"
  },
  {
   "time": "2023-01-17 08:06:00 UTC"
  },
  {
   "time": "2023-01-17 08:12:00 UTC"
  },
  {
   "time": "2023-01-17 08:18:00 UTC"
  },
  {
   "time": "2023-01-17 08:24:00 UTC"
  },
  {
   "time": "2023-01-17 08:30:00 UTC"
  },
  {
   "time": "2023-01-17 08:36:00 UTC"
  },
  {
   "time": "2023-01-17 08:42:00 UTC"
  },
  {
   "time": "2023-01-17 08:48:00 UTC"
  },
  {
   "time": "2023-01-17 08:54:00 UTC"
  },
  {
   "time": "2023-01-17 09:00:00 UTC"
  },
  {
   "time": "2023-01-17 09:06:00 UTC"
  },
  {
   "time": "2023-01-17 09:12:00 UTC"
  },
  {
   "time": "2023-01-17 09:18:00 UTC"
  },
  {
   "time": "2023-01-17 09:24:00 UTC"
  },
  {
   "time": "2023-01-17 09:30:00 UTC"
  },
  {
   "time": "2023-01-17 09:36:00 UTC"
  },
  {
   "time": "2023-01-17 09:42:00 UTC"
  },
  {
   "time": "2023-01-17 09:48:00 UTC"
  },
  {
   "time": "2023-01-17 09:54:00 UTC"
  },
  {
   "time": "2023-01-17 10:00:00 UTC"
  },
  {
   "time": "2023-01-17 10:06:00 UTC"
  },
  {
   "time": "2023-01-17 10:12:00 UTC"
  },
  {
   "time": "2023-01-17 10:18:00 UTC"
  },
  {
   "time": "2023-01-17 10:24:00 UTC"
  },
  {
   "time": "2023-01-17 10:30:00 UTC"
  },
  {
   "time": "2023-01-17 10:36:00 UTC"
  },
  {
   "time": "2023-01-17 10:42:00 UTC"
  },
  {
   "time": "2023-01-17 10:48:00 UTC"
  },
  {
   "time": "2023-01-17 10:54:00 UTC"
  },
  {
   "time": "2023-01-17 11:00:00 UTC"
  },
  {
   "time": "2023-01-17 11:06:00 UTC"
  },
  {
   "time": "2023-01-17 11:12:00 UTC"
  },
  {
   "time": "2023-01-17 11:18:00 UTC"
  },
  {
   "time": "2023-01-17 11:24:00 UTC"
  },
  {
   "time": "2023-01-17 11:30:00 UTC"
  },
  {
   "time": "2023-01-17 11:36:00 UTC"
  },
  {
   "time": "2023-01-17 11:42:00 UTC"
  },
  {
   "time": "2023-01-17 11:48:00 UTC"
  },
  {
   "time": "2023-01-17 11:54:00 UTC"
  },
  {
   "time": "2023-01-17 12:00:00 UTC"
  },
  {
   "time": "2023-01-17 12:06:00 UTC"
  },
  {
   "time": "2023-01-17 12:12:00 UTC"
  },
  {
   "time": "2023-01-17 12:18:00 UTC"
  },
  {
   "time": "2023-01-17 12:24:00 UTC"
  },
  {
   "time": "2023-01-17 12:30:00 UTC"
  },
  {
   "time": "2023-01-17 12:36:00 UTC"
  },
  {
   "time": "2023-01-17 12:42:00 UTC"
  },
  {
   "time": "2023-01-17 12:48:00 UTC"
  },
  {
   "time": "2023-01-17 12:54:00 UTC"
  },
  {
   "time": "2023-01-17 13:00:00 UTC"
  },
  {
   "time": "2023-01-17 13:06:00 UTC"
  },
  {
   "time": "2023-01-17 13:12:00 UTC"
  },
  {
   "time": "2023-01-17 13:18:00 UTC"
  },
  {
   "time": "2023-01-17 13:24:00 UTC"
  },
  {
   "time": "2023-01-17 13:30:00 UTC"
  },
  {
   "time": "2023-01-17 13:36:00 UTC"
  },
  {
   "time": "2023-01-17 13:42:00 UTC"
  },
  {
   "time": "2023-01-17 13:48:00 UTC"
  },
  {
   "time": "2023-01-17 13:54:00 UTC"
  },
  {
   "time": "2023-01-17 14:00:00 UTC"
  },
  {
   "time": "2023-01-17 14:06:00 UTC"
  },
  {
   "time": "2023-01-17 14:12:00 UTC"
  },
  {
   "time": "2023-01-17 14:18:00 UTC"
  },
  {
   "time": "2023-01-17 14:24:00 UTC"
  },
  {
   "time": "2023-01-17 14:30:00 UTC"
  },
  {
   "time": "2023-01-17 14:36:00 UTC"
  },
  {
   "time": "2023-01-17 14:42:00 UTC"
  },
  {
   "time": "2023-01-17 14:48:00 UTC"
  },
  {
   "time": "2023-01-17 14:54:00 UTC"
  },
  {
   "time": "2023-01-17 15:00:00 UTC"
  },
  {
   "time": "2023-01-17 15:06:00 UTC"
  },
  {
   "time": "2023-01-17 15:12:00 UTC"
  },
  {
   "time": "2023-01-17 15:18:00 UTC",
   "is_partial": true
  }
 ]
}
```

Reply:
Compared to master:
```

PS C:\Users\Work\Documents\GitHub\gocryptotrader\cmd\gctcli> go run . gethistoriccandlesextended --exchange="binance" --pair="btc-usdt" --asset="spot" -i="60" --start="2023-01-17 13:01:00" --end="2023-01-18 02:30:00"
{
 "exchange": "Binance",
 "pair": {
  "delimiter": "-",
  "base": "btc",
  "quote": "usdt"
 },
 "start": "2023-01-17 13:01:00 AEDT",
 "end": "2023-01-18 02:30:00 AEDT",
 "interval": "1m",
 "candle": [
  {
   "time": "2023-01-17 02:01:00 UTC",
   "low": 21048.27,
   "high": 21064.56,
   "open": 21057.95,
   "close": 21051.57,
   "volume": 151.5093
  },

  ... snipped for sanity

  {
   "time": "2023-01-17 03:05:00 UTC",
   "low": 21092.06,
   "high": 21093.33,
   "open": 21092.06,
   "close": 21092.34,
   "volume": 1.25656
  }
 ]
}
```
So it cuts off times that don't yet exist, but doesn't complain either. So I'm happy for a trim instead of error